### ISSUE TITLE
Mutating objects when index not available

### ISSUE THREAD

**Issue by yarnball** [kolodny/immutability-helper#53]
Hi,

I am having trouble using the `immutability-helper` on data as the index it gets is incorrect.

I have created a side-by-side example, one uses `state`, and the other uses css to hide values (and thus the index remains the same.

How can I update/modify items when the index is not available?

Thanks.

I put this in a codeSandbox to:
https://codesandbox.io/s/o2rDJZ7pz

```
class App extends React.Component {
  ...
  remove = (e, index) => {
    this.setState({
      movies: update(this.state.movies, { $splice: [[index, 1]] }),
    });
  };
  ....
            {moviesToHide.map((x, index) => {
              return (
                <ResultState
                  key={x.id}
                  result={x}
                  index={index}
                  remove={e => this.remove(e, index)}
                />
              );
            })}
          </div>
```

**Comment by kolodny** [comment]
It's a little hard to follow what's happening here, can you make a smaller example of what's breaking?

BTW you don't need to do this:
```js
this.setState({
  movies: update(this.state.movies, { $splice: [[index, 1]] }),
});
```

You can do this instead:

```js
this.setState(update(state, {
  movies: { $splice: [[index, 1]] },
});

```

**Comment by yarnball** [comment]
I'll explain a little better. I can use that "del" button to remove my item using index. However, when I filter the state `moviesToHide`, it then re-indexes all the items.

For example:

- Value `Arcu tempor` is `index 22`, `Elit et` is at `index 0`.
- I search for `Arcu tempor` **becomes** `index 0`. 
- I use `movies: { $splice: [[index, 1]] },` 
- It removes the `Elit et`, even though `Arcu tempor` shows at `index 0`.

I've compacted my example. Check the codeSandbox to see it in action https://codesandbox.io/s/o2rDJZ7pz .


```
class App extends React.Component {
  state = {
    movies: largeData.movies,
    searchQuery: '',
  };
  remove = (e, index) => {
    console.log('removed index:', index);
    this.setState(
      update(this.state, {
        movies: { $splice: [[index, 1]] },
      }),
    );
  };
  render() {
    const { searchQuery, movies } = this.state;
    const moviesToHide = filterMovies(movies, searchQuery.split(' '));
    var moviesNamesToHide = new Set(moviesToHide.map(m => m.name));

    return (
      <div>
        Search content
        <input
          placeholder="type Q"
          onChange={e => this.setState({ searchQuery: e.target.value })}
        />
            {moviesToHide.map((x, index) => {
              return (
                <ResultState
                  key={x.id}
                  result={x}
                  index={index}
                  remove={e => this.remove(e, index)}
                />
              );
            })}
      </div>
    );
  }
}
```

**Comment by kolodny** [comment]
I see. The problem is that you're removing the element at the index of the filtered item instead of the index of the item in `this.state.movies`. Here's a fixed version. https://codesandbox.io/s/Rgm8Njryw . The main difference is the following:

```diff
- remove = (e, index) => {
+ remove = (e, movie) => {
+   const index = this.state.movies.indexOf(movie);
    console.log('removed index:', index);
    this.setState(
      update(this.state, {
        movies: { $splice: [[index, 1]] },
      }),
    );
  };

- remove={e => this.remove(e, index)}
+ remove={e => this.remove(e, x)}
```
